proposals/nnnn-actors.md

```diff
@@ -140,6 +140,265 @@ extension BankAccount {
+
+Currently, this proposal takes the aggressive approach of assuming _all_ actors are reentrant and not providing ways to opt out of this behavior. This section aims to highlight issues, benefits and tradeoffs with this approach as well as non-reentrancy in general. The goal of discussing these reentrancy issues, is to arrive at a design that is developer friendly, non-surprising, and delivers more completely on Swift's promise of _safe_ and _sane_ concurrency by default thanks to the use of actors.
+
+#### Reentrant actors: "Interleaving" execution
+
+Reentrancy means that execution of asynchronous actor-isolated functions may "interleave" at suspension points, leading to increased complexity in programming with such actors, as every suspension point must be carefully inspected if the code _after_ it depends on some invariants that could have changed before it suspended.
+
+Interleaving executions still respect the actor's "single-threaded illusion"–i.e. no two functions will ever execute *concurrently* on any given actor–however they may _interleave_ at suspension points. In broad terms this means that reentrant actors are "thread-safe" but are not "data race safe," meaning it is not possible to observe any
+
+> Empirically: we know that both an non-reentrant and reentrant awaiting and actors are useful, however both semantics must be available to developers in order to use actors as a means of isolating state from "concurrent" (in the meaning of interleaved) modification.
+
+To further clarify the implications of this, let us consider the following actor, which thinks of an idea and then returns it, after telling its friend about it.
+
+```swift
+// reentrant actor
+actor class Person {
+  let friend: Friend
+  
+  // actor-isolated opinion
+  var opinion: Judgement = .noIdea
+
+  func thinkOfGoodIdea() async -> Decision {
+    opinion = .goodIdea                    // <1>
+    await tell(friend, myOpinion: opinion) // <2>
+    return opinion // 🤨                   // <3>
+  }
+
+  func thinkOfBadIdea() async -> Decision {
+    opinion = .badIdea                     // <4>
+    await tell(friend, myOpinion: opinion) // <4>
+    return opinion // 🤨                   // <6>
+  }
+}
+```
+
+> Reentrant code is notoriously difficult to program with–one is protected from low level data-races, however the higher level semantic races may still happen. In the examples shown, the states are small enough that they are simple to fit in one's hear so they do not appear as tricky, however in the real world races can be dauntingly hard to debug.
+
+In the example above the `Person` can think of a good or bad idea, shares that opinion with a friend, and returns that opinion that it stored. Since the actor is reentrant this code is **wrong** and will return an _arbitrary opinion_ if the actor begins to think of a few ideas at the same time.
```

Comment:
I mean, this is really weird code.  Why is `opinion` actor state to begin with?

Reply:
The example is silly / oversimplified to easily spot the issue -- i can make some bigger examples where it'd show up as suprising; e.g. using some `self.peers` in multiple steps of a quorum calculation, while in the middle of doing so someone might have reentrantly changed our set of peers etc.

Reply:
A more realistic example perhaps:

```
actor Acceptor { 
var peers: [Peer]

  func quorum(...) async -> Bool { 
    let oks: [Prepare] = []
    for peer in peers {
      await oks.append(peer.prepare(...))
    }
    log.info("Got OKs:\(oks) from \(peers.count)") // could be wrong "wrong" peers count

    guard await enough(oks) else { return false }

    for peer in peers {
      await peer.accept(...)
    }

    // implicit assumption that peers did not change between the runs?
  }
}
```

One would need to copy the peers to function scope to prevent them from accidentally be changed I guess...

It's a bit fuzzy to find a great example... should I try to find better ones?